src/axom/core/Array.hpp

```diff
@@ -1364,7 +1364,15 @@ inline void Array<T, DIM, SPACE>::dynamicRealloc(IndexType new_num_elements)
     utilities::processAbort();
   }
 
-  m_data = axom::reallocate<T>(m_data, new_capacity, m_allocator_id);
+  if(m_num_elements == 0)
+  {
+    axom::deallocate(m_data);
+    m_data = axom::allocate<T>(new_capacity, m_allocator_id);
+  }
+  else
+  {
+    m_data = axom::reallocate<T>(m_data, new_capacity, m_allocator_id);
+  }
```

Comment:
This was changed to fix a memory error in `spin_uniform_grid_test` in corona gitlab CI and occurred sporadically when I self ran it:
https://github.com/LLNL/axom/blob/cf9ae83a827b6a568f797459e45cb9f3ffd4e893/src/axom/spin/tests/spin_uniform_grid.cpp#L51-L56

The error occurs when resizing axom::Array `m_bins` from size 0 to 1000000 (?):
https://github.com/LLNL/axom/blob/cf9ae83a827b6a568f797459e45cb9f3ffd4e893/src/axom/spin/policy/UniformGridStorage.hpp#L53

Umpire throws an exception for `reallocate()`, but not a new `allocate()`.

@kanye-quest, I'm wary of adding this change to `axom::Array`. Thoughts on 
what could be the cause?

Backtrace of the error:
[spin_uniform_grid_backtrace.txt](https://github.com/LLNL/axom/files/8994739/spin_uniform_grid_backtrace.txt)

Reply:
@bmhan12 , I agree that this seems strange and should be resolved before committing.

Reply:
I tried playing around with this a little while ago -- it seems like we'll sometimes get junk addresses with `malloc/realloc()`:

When the test passes:
```
[ RUN      ] spin_uniform_grid.indexing
[Instance 0x7fffffffbc80] Pointer address before realloc: 0x0
[Instance 0x7fffffffbc80] Pointer address after realloc: 0x660ebe0
[Instance 0x7fffffffbc80] Pointer address before realloc: 0x660ebe0
[Instance 0x7fffffffbc80] Pointer address after realloc: 0x660f4c0
``` 

When the test fails:
```
[ RUN      ] spin_uniform_grid.indexing
[Instance 0x7fffffffbc80] Pointer address before realloc: 0x0
[Instance 0x7fffffffbc80] Pointer address after realloc: 0x15554703d760
[Instance 0x7fffffffbc80] Pointer address before realloc: 0x15554703d760
--- segfaults here ---
```

FWIW, it doesn't seem like this problem appears in rznevada. On corona, our `malloc/realloc` seems to come from `mvapich2-2.3.6-aocc-4.5.2.0/lib/libmpi.so.12`, while on rznevada we're using `cray-mpich-8.1.16-rocmcc-5.1.1`. Maybe a bug in corona's MPI library?

Reply:
Can we switch our CI from `corona` to `tioga` as part of this PR? Or perhaps as a standalone PR that merges before this one?

We've been getting a number of system-related failures on `corona` in the past few weeks, and we're much more interested in the behavior on `tioga`.

Reply:
@kennyweiss  I think switching to tioga is a good idea, that can be done in a standalone PR. I am concerned with 
tioga/rzvernal currently being unstable, with nodes permanently going down on segfaults and requiring a restart by LC.  A fix looks to be coming today, so we'll have to see...

I will open a separate PR for tioga, once the tioga update goes through and gitlab CI comes back online to verify it works.

Reply:
@kanye-quest  In the meantime, I've bypassed this error on `corona ` and reverting this axom::Array change by initializing the Uniform Grid in unified memory.

Reply:
@kennyweiss  Yes, we could definitely do that.